compute_tools/src/vm_monitor/cgroup.rs

```diff
@@ -0,0 +1,777 @@
+
+    /// The sequence number of the last upscale.
+    ///
+    /// If we receive a memory.high event that has a _lower_ sequence number than
+    /// `last_upscale_seqnum`, then we know it occured before the upscale, and we
+    /// can safely ignore it.
+    ///
+    /// Note: Like the `events` field, this doesn't _need_ interior mutability but we
+    /// use it anyways so that methods take `&self`, not `&mut self`.
+    last_upscale_seqnum: AtomicU64,
+
+    /// The name of the cgroup we are managing
+    // Won't be necessary if/once https://github.com/kata-containers/cgroups-rs/pull/120
+    // is merged
+    name: String,
+
+    /// A channel on which we send messages to request upscale from the dispatcher.
+    upscale_requester: Sender<()>,
+
+    /// The actual cgroup we are watching and managing.
+    cgroup: cgroups_rs::Cgroup,
+
+    /// The direct source of upscale events, for if we need to listen for an
+    /// upscale event _in particular_.
+    ///
+    /// All events from this channel are also piped into `self.events`. Thus,
+    /// when awaiting upscales directly, first check if an upscale has already
+    /// been peeked and is held in events, then directly await this channel.
+    upscale_receiver: Receiver<Sequenced<Resources>>,
+}
+
+/// Stream of events that [`CgroupWatcher::watch`] takes as input
+pub struct CgroupWatcherEventStream {
+    // We're using a full type to wrap the stream so that we dno't need to make the inner types
+    // public.
+    stream: PeekableStream<Sequenced<EventKind>>,
+}
+
+/// Read memory.events for the desired event type.
+///
+/// `path` specifies the path to the desired `memory.events` file.
+/// For more info, see the `memory.events` section of the [kernel docs]
+/// <https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-interface-files>
+fn get_event_count(path: &str, event: MemoryEvent) -> anyhow::Result<u64> {
+    let contents = fs::read_to_string(path)
+        .with_context(|| format!("failed to read memory.events from {path}"))?;
+
+    // Then contents of the file look like:
+    // low 42
+    // high 101
+    // ...
+    contents
+        .lines()
+        .filter_map(|s| s.split_once(' '))
+        .find(|(e, _)| *e == event.as_str())
+        .ok_or_else(|| anyhow!("failed to find entry for memory.{event} events in {path}"))
+        .and_then(|(_, count)| {
+            count
+                .parse::<u64>()
+                .with_context(|| format!("failed to parse memory.{event} as u64"))
+        })
+}
+
+/// Create an event stream that produces events whenever the file at the provided
+/// path is modified.
+fn create_file_watcher(path: &str) -> anyhow::Result<EventStream<[u8; 1024]>> {
+    info!("creating file watcher for {path}");
+    let inotify = Inotify::init().context("failed to initialize file watcher")?;
+    inotify
+        .watches()
+        .add(path, WatchMask::MODIFY)
+        .with_context(|| format!("failed to start watching {path}"))?;
```

Comment:
I do recall us already using some wrapper which supports linux and mac but this does not look like it. Mac support is in general needed (unless we've dropped this recently).

With the inotify wrapper we fall back to polling the file ... Maybe it was the `postmaster.pid`, though could be that implementation is long gone.

Reply:
This feature inherently isn't supported on mac because it uses cgroups, so I'll keep using `inotify` as it's a bit simpler. However, this reminded me that I need to add some `#[cfg]` directives so I don't break all mac builds (`cgroup-rs` doesn't compile on non-linux platforms).